### PR TITLE
[Draft] Add LIFX Switch support

### PR DESCRIPTION
**Pull request by Samywamy10**
Draft for adding Switch support. This doesn't work (or at least I can't get it to work) but should be somewhat of a basis...

Sorry for the random PR - but I'm stuck with an issue:

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/homeassistant/helpers/update_coordinator.py", line 239, in _async_refresh
    self.data = await self._async_update_data()
  File "/config/custom_components/lifx/coordinator.py", line 240, in _async_update_data
    await self.async_get_rpower(None)
  File "/config/custom_components/lifx/coordinator.py", line 493, in async_get_rpower
    await async_execute_lifx(partial(self.device.get_rpower, relay_index))
AttributeError: 'Light' object has no attribute 'get_rpower'
```

I'm introducing the `get_rpower` method in aiolifx [here](https://github.com/frawau/aiolifx/pull/67), and pointing to it in the  Poetry deps list. But when I test in HA, I guess it doesn't have the version I'm pointing to. Wondering if I'd be able to get some help please? (short of waiting for the PR to `aiolifx` to be merged)

### CODE REVIEW

**Comment by Djelibeybi** [comment]
The LIFX Switch is supported by the HomeKit Integration a lot better than it can be via the LAN protocol (or HTTP API) as neither expose the button press events, just the terminals. It may be possible to add support for terminals connected to non-LIFX devices, though.

I've not actually tried the `aiolifx` implementation PR but it works using Photons.

**Comment by Djelibeybi** [comment]
I'll try and find some time over the next day or two to test our your PR though.

**Comment by Samywamy10** [comment]
> The LIFX Switch is supported by the HomeKit Integration a lot better than it can be via the LAN protocol (or HTTP API) as neither expose the button press events, just the terminals. It may be possible to add support for terminals connected to non-LIFX devices, though.

Yeah; I may end up going down that route but the immediate problem I wanted to solve was being able to turn on the relays for non-LFIX devices as you say.

**Comment by Djelibeybi** [comment]
If you want local control, but don't want to use HomeKit Controller, you could use Photons Interactor and Node-RED: https://community.lifx.com/t/home-assistant-lifx-switch/7378/18?u=djelibeybi

This can also be adjusted to use native RESTful Switches in Home Assistant, though I've not actually done that.

I did spend a bunch of time testing the HomeKit Controller integration with the LIFX firmware engineers and it works really well. You only need an iOS device to get the HomeKit code (if your Switch is old enough not to have come with one). All other config can be done directly on Home Assistant.

I will be upfront and say I'm not going to merge anything that doesn't use a released version of `aiolifx` so until that PR is merged, this PR is probably going to be either. But I'm happ7y to continue trying to help you achieve what you need.

**Comment by Samywamy10** [comment]
> I will be upfront and say I'm not going to merge anything that doesn't use a released version of aiolifx so until that PR is merged, this PR is probably going to be either. But I'm happy to continue trying to help you achieve what you need.

All good I'm not expecting this to be merged yet; just wanted to try to get it working. Regardless, my PR has been merged and version bumped of aiolifx, I think it just needs a new release to be created?

**Comment by Djelibeybi** [comment]
To be honest, I haven't taken a look at how this is implemented in `aiolifx` but I wouldn't add the Switch functions to the Light class. 

Rather, I would have created a new Switch class derived from Device so that it can also handle the [StateUnhandled](https://lan.developer.lifx.com/docs/information-messages#stateunhandled---packet-223) response.

**Comment by Samywamy10** [comment]
I've taken some notes of things I'd like to do in follow up PRs. It would have made it a bigger diff (imo beyond the scope of "supportinga switch") as atm every device is instantiated as a `Light` (as IIUC during discovery we don't know the product ID, it needs to be looked up as another message) instead of a `Device` and then narrowed later to a `Light`.

Open to ideas:
- add switch: https://github.com/frawau/aiolifx/pull/67/files
- small refactorings: https://github.com/frawau/aiolifx/pull/68

**Comment by Djelibeybi** [comment]
We worked around this in Home Assistant by creating the LIFXConnection class which creates the initial UDP protocol and to which we can send packets. Once we have the product ID, we know the type of device and once we have the firmware version, we know whether the serial number is the same as the MAC address or not.

BTW, what `aiolifx` refers to as `mac_addr` is not the MAC address all the time. It's always the serial (without the colons), but if the device is running firmware 3,70 or higher, the mac address is off by one in the last octet.

**Comment by Samywamy10** [comment]
Thanks that's some useful info. I've linked to your comments in [the issue](https://github.com/frawau/aiolifx/issues/69) I've created in aiolifx to track this.

Once the aiolifx version is bumped I'll come back to this PR as I'm still not sure how to get my own install of HA to use my branched version of aiolifx.

**Comment by Samywamy10** [comment]
Hey, I've mucked around with this a bit now and it works if you'd like to test. It looks like `discovery.label` returns a different label to `light.label` (which is undefined) so when you go to add the entity its unnamed. Not sure how to get access to this. There's also a bit of a delay when you turn on the switch and the UI responds.

Open to ideas/PR comments given it works now

**Comment by Djelibeybi** [comment]
I haven't had a chance to play with this yet, unfortunately. I'm also planning on archiving this repo shortly as I've decided to to stop being a maintainer of the core integration. Eventually, I may release my Photons-based integration but it'll be a while before that happens (if at all).

I encourage you to fork this repo and publish and maintain your own version, though. :)

File: custom_components/lifx/coordinator.py [hunk 1]
```diff
@@ -180,6 +180,13 @@ async def diagnostics(self) -> dict[str, Any]:
         if features["infrared"] is True:
             device_data["infrared"] = {"brightness": self.device.infrared_brightness}
 
+        if features["relays"] is True:
+            device_data["relays"] = {
+                "Relay 1": self.device.relays_power[0],
+                "Relay 2": self.device.relays_power[1],
+                "Relay 3": self.device.relays_power[2],
+                "Relay 4": self.device.relays_power[3]
```
Comment (Djelibeybi) [inline custom_components/lifx/coordinator.py:188]: The problem with this (and the Switch API in general) is that this will create two extra relays on 2-gang switches. 

Also, while the API calls them relays, the LIFX app calls them terminals. I suggest calling them "Terminal 1", "Terminal 2", etc instead as this matches the app and will make it easier to document.

File: custom_components/lifx/light.py [hunk 2]
```diff
@@ -95,6 +95,10 @@ async def async_setup_entry(
         LIFX_SET_HEV_CYCLE_STATE_SCHEMA,
         "set_hev_cycle_state",
     )
+
+    if lifx_features(device)["relays"]:
+        return
```
Comment (Djelibeybi) [inline custom_components/lifx/light.py:100]: Why is this here? The other `if` statements would limit the creation to Lights anyway (as each tests for a specific light capability).

File: custom_components/lifx/switch.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up LIFX from a config entry."""
+    domain_data = hass.data[DOMAIN]
+    coordinator: LIFXUpdateCoordinator = domain_data[entry.entry_id]
+    if lifx_features(coordinator.device)["relays"]:
+        async_add_entities([LIFXSwitch(coordinator, LIFX_SWITCH, 0), LIFXSwitch(coordinator, LIFX_SWITCH, 1), LIFXSwitch(coordinator, LIFX_SWITCH, 2), LIFXSwitch(coordinator, LIFX_SWITCH, 3)])
```
Comment (Djelibeybi) [inline custom_components/lifx/switch.py:29]: What about 2-gang switches? Unfortunately, there's no way to tell from the API, so you may have to create a config flow for the switch.

File: custom_components/lifx/switch.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    def __init__(self, coordinator: LIFXUpdateCoordinator, description: SwitchEntityDescription, relay_index: int) -> None:
+        super().__init__(coordinator)
+        self.relay_index = relay_index
+        self._attr_unique_id = f"{coordinator.serial_number}_{description.key}_{relay_index + 1}"
+        self._attr_name = f"Relay {relay_index + 1}"
+        self.entity_description = description
+    
+    @property
+    def is_on(self) -> bool:
+        return self.bulb.relays_power[self.relay_index]
```
Comment (Djelibeybi) [inline custom_components/lifx/switch.py:44]: This really should be `self.device`. Having `self.bulb` for a switch is weird. :)

**Comment by Djelibeybi** [comment]
My comments/feedback are just from reading the code. I haven't actually tested it yet. Also, I marked them as comments, not requested changes.

**Comment by ironcrafter54** [comment]
are there any plans to continue this in the future?

**Comment by Djelibeybi** [comment]
Not at the moment. I meant to archive the repo ages ago but forgot. I'll do that now, so it's explicit.
